### PR TITLE
buildGoPackage: use $out instead of $bin

### PR DESCRIPTION
**Pull request by zowoq**
cc @kalbasit @Mic92

This changes `buildGoPackage` to use to `$out` instead of `$bin` which matches the behaviour of `buildGoModule` (and `buildRustPackage`).

https://github.com/NixOS/nixpkgs/blob/a2d9f5843336b00398542448321a497309907d05/pkgs/servers/sql/cockroachdb/default.nix#L53-L57

### CODE REVIEW

**Comment by adisbladis** [comment]
In https://github.com/NixOS/nixpkgs/pull/43328 I opted _not_  to do this as it would introduce an incompatibility for questionable gain.

**Comment by kalbasit** [comment]
Can we alias out to bin for backward compatibility?

**Comment by Mic92** [comment]
> Can we alias out to bin for backward compatibility?

It could be a symlink I think. However I suspect most of the breakages occur in the build expressions itself because `$bin` needs to be changed to `$out`.

File: pkgs/development/go-packages/generic/default.nix [hunk 1]
```diff
@@ -245,7 +245,7 @@ let
     } // meta // {
       # add an extra maintainer to every package
       maintainers = (meta.maintainers or []) ++
-                    [ lib.maintainers.lethalman ];
+                    [ ];
```
Comment (Mic92) [inline pkgs/development/go-packages/generic/default.nix:245]: Just drop the whole `// { maintainers = ...; }`.

**Comment by kalbasit** [comment]
How about deprecate the output (not sure that's possible, at least the trace part), spread the word to give people a couple of months before we do this? In the meantime we could gate the output change with an attr and set it from within nixpkgs.

**Comment by zowoq** [comment]
I might do another round of `$bin` cleanup but starting to get to the point of diminishing returns.

I'm not sure how to move forward with this?

**Comment by Mic92** [comment]
The `mkdir` in `nixpkgs/pkgs/servers/monitoring/uchiwa/default.nix` needs a `-p`.

**Comment by zowoq** [comment]
It already has `-p`?

File: pkgs/servers/monitoring/uchiwa/default.nix [hunk 2]
```diff
@@ -37,7 +37,7 @@ in stdenv.mkDerivation {
 
   buildCommand = ''
     mkdir -p $out/bin $out/public
```
Comment (zowoq) [inline pkgs/servers/monitoring/uchiwa/default.nix:39]: ?

**Comment by zowoq** [comment]
The other `mkdir`. Fixed.

**Comment by Mic92** [comment]
Could you also write a short entry in `doc/manual/release-notes/rl-2009.xml`. This is a breaking change for all out-of-tree `buildGoPackages`.

**Comment by zowoq** [comment]
Added a note and dropped a paragraph from the go docs that referenced defaulting to `$bin`.

**Comment by Mic92** [comment]
Merged into staging in 20e67f1fb83b2a4a43571d9a648dee07e2cd15af

**Pull request closed without merging**